### PR TITLE
Fixes #10050: Remove all datasource code from Rudder main and add needed hooks

### PR DESCRIPTION
**Pull request by fanf**
https://www.rudder-project.org/redmine/issues/10050

### CODE REVIEW

**Comment by fanf** [comment]
PR rebased

File: rudder-core/src/main/resources/Migration/dbMigration-4.0.x-4.1-add-datasource-table.sql [hunk 1]
```diff
@@ -38,6 +38,9 @@
 -- Create the table for the node configuration
 CREATE TABLE dataSources (
```
Comment (VinceMacBuche) [inline rudder-core/src/main/resources/Migration/dbMigration-4.0.x-4.1-add-datasource-table.sql]: That file can be deleted (and removed from migration script)

Comment (fanf) [inline rudder-core/src/main/resources/Migration/dbMigration-4.0.x-4.1-add-datasource-table.sql]: I did deleted it, but certainly forgot to commit.

**Comment by fanf** [comment]
Not to merge now: the UI part is not removed. And it would be better to commit the UIv2 before removing (if ready this morning).

**Comment by fanf** [comment]
PR rebased

File: rudder-core/src/main/resources/Migration/dbMigration-4.0.x-4.1-add-datasource-table.sql (hunk 1 above)
Comment (VinceMacBuche) [inline rudder-core/src/main/resources/Migration/dbMigration-4.0.x-4.1-add-datasource-table.sql]: Still here :D

**Comment by fanf** [comment]
PR rebased

File: rudder-core/src/main/scala/com/normation/rudder/domain/nodes/Node.scala [hunk 2]
```diff
@@ -135,22 +180,41 @@ object CompareProperties {
    *   the property
    * - if the value is the emtpy string, remove
    *   the property
+   *
+   * Each time, we have to check the provider of the update to see if it's compatible.
+   * Node that in read-write mode, the provider is the last who wrote the property.
    */
-  def updateProperties(props: Seq[NodeProperty], updates: Option[Seq[NodeProperty]]) = {
+  def updateProperties(props: Seq[NodeProperty], updates: Option[Seq[NodeProperty]], provider: NodePropertyProvider): Box[Seq[NodeProperty]] = {
```
Comment (VinceMacBuche) [inline rudder-core/src/main/scala/com/normation/rudder/domain/nodes/Node.scala]: I don't think the provider (third parameters) here is used

File: rudder-core/src/main/scala/com/normation/rudder/domain/nodes/Node.scala [hunk 3]
```diff
@@ -135,22 +180,41 @@ object CompareProperties {
...
+    import NodePropertyMode._
     updates match {
-      case None => props
+      case None => Full(props)
       case Some(u) =>
-        val values = u.map { case NodeProperty(k, v) => (k, v)}.toMap
+        val values = u.map { case NodeProperty(k, v, p, m) => (k, (v, p, m))}.toMap
```
Comment (VinceMacBuche) [inline rudder-core/src/main/scala/com/normation/rudder/domain/nodes/Node.scala]: Maybe we could have done a Map[String,NodeProperty] here ?

u.map(p => (p.name,p)

File: rudder-core/src/main/scala/com/normation/rudder/domain/nodes/Node.scala [hunk 4]
```diff
@@ -90,7 +94,46 @@ case object Node {
...
+ *
+ * If mode is read-only, only the provider can modify
+ * the properties. This is typically what happens
+ * when the property is set by an external source.
+ */
+final case class NodeProperty(
+    name    : String
+  , value   : JValue
+  , provider: Option[NodePropertyProvider] // optional, default "rudder"
+  , mode    : Option[NodePropertyMode]     // optional, default, "read-write"
```
Comment (VinceMacBuche) [inline rudder-core/src/main/scala/com/normation/rudder/domain/nodes/Node.scala]: You may need to change "mode" by  "access" or "permissions"

File: rudder-core/src/main/scala/com/normation/rudder/domain/nodes/Node.scala [hunk 5]
```diff
@@ -135,22 +180,41 @@ object CompareProperties {
...
+        //for news values, don't keep empty, that make no sense
+        val news = (values -- existings).collect { case(k, (v, p, m)) if(v != JString("")) => NodeProperty(k,v,p,m) }
+
+        //for old value, update only according to rights
+        for {
+          updated <- sequence(props) { case prop@NodeProperty(name, value, provider, mode)  =>
+                       values.get(name) match {
+                         case None            => Full(Some(prop))
+                         case Some((v, p, m)) =>
+                           (mode, provider, p) match {
```
Comment (VinceMacBuche) [inline rudder-core/src/main/scala/com/normation/rudder/domain/nodes/Node.scala]: Do we need to match on provider and p ? 

I think you only need to match on 'mode' and not on providers ? 

(first case will be 

case Some(ReadOnly)  if provider == p

File: rudder-core/src/main/scala/com/normation/rudder/services/policies/DeploymentService.scala [hunk 6]
```diff
@@ -129,11 +147,10 @@ trait PromiseGenerationService extends Loggable {
...
 
-      // updating node properties from data source must be node before getting node info
-      fetchDatasourcesTime =  System.currentTimeMillis
-      _                    =  datasourceCallbacks.onGenerationStarted(generationTime)
-      timeFetchDatasources =  (System.currentTimeMillis - fetchDatasourcesTime)
-      _                    =  logger.debug(s"All data fetched from data source in ${timeFetchDatasources} ms, start getting all generation related data.")
+      codePreGenHooksTime =  System.currentTimeMillis
+      _                   <- beforeDeploymentSync(generationTime)
+      timeCodePreGenHooks =  (System.currentTimeMillis - codePreGenHooksTime)
+      _                   =  logger.debug(s"All data fetched from data source in ${timeCodePreGenHooks} ms, start getting all generation related data.")
```
Comment (VinceMacBuche) [inline rudder-core/src/main/scala/com/normation/rudder/services/policies/DeploymentService.scala]: You may want to change the text here to "All pre generation hooks executed in"

File: rudder-core/src/main/scala/com/normation/rudder/services/policies/DeploymentService.scala [hunk 7]
```diff
@@ -249,7 +266,8 @@ trait PromiseGenerationService extends Loggable {
 
     } yield {
       logger.debug("Timing summary:")
-      logger.debug("Run pre generation hooks  : %10s ms".format(timeRunPreGenHooks))
+      logger.debug("Run pre-gen scripts hooks : %10s ms".format(timeRunPreGenHooks))
+      logger.debug("Run pre-gen code hooks     : %10s ms".format(timeCodePreGenHooks))
```
Comment (VinceMacBuche) [inline rudder-core/src/main/scala/com/normation/rudder/services/policies/DeploymentService.scala]: Can you remove a space before the colon so it's correctly aligned ??

File: rudder-web/src/main/scala/bootstrap/liftweb/AppConfig.scala [hunk 8]
```diff
@@ -1527,7 +1512,7 @@ object RudderConfig extends Loggable {
   private[this] lazy val reportsRepositoryImpl = new ReportsJdbcRepository(doobie)
   private[this] lazy val complianceRepositoryImpl = new ComplianceJdbcRepository(doobie)
   private[this] lazy val dataSourceProvider = new RudderDatasourceProvider(RUDDER_JDBC_DRIVER, RUDDER_JDBC_URL, RUDDER_JDBC_USERNAME, RUDDER_JDBC_PASSWORD, RUDDER_JDBC_MAX_POOL_SIZE)
-  private[this] lazy val doobie = new Doobie(dataSourceProvider.datasource)
+  lazy val doobie = new Doobie(dataSourceProvider.datasource)
```
Comment (VinceMacBuche) [inline rudder-web/src/main/scala/bootstrap/liftweb/AppConfig.scala]: I was quitte puzzled by the dataSourceProvider (from doobie)

File: rudder-core/src/main/scala/com/normation/rudder/domain/nodes/Node.scala (hunk 2 above)
Comment (fanf) [inline rudder-core/src/main/scala/com/normation/rudder/domain/nodes/Node.scala]: it's used in the comparison to check in the read-only case: the same provider can modify, whatever the access mode.

File: rudder-core/src/main/scala/com/normation/rudder/domain/nodes/Node.scala (hunk 3 above)
Comment (fanf) [inline rudder-core/src/main/scala/com/normation/rudder/domain/nodes/Node.scala]: why not, effectively

File: rudder-core/src/main/scala/com/normation/rudder/domain/nodes/Node.scala [hunk 9]
```diff
@@ -135,22 +180,41 @@ object CompareProperties {
...
-            case Some(JString("")) => None
-            case Some(x)           => Some(NodeProperty(name, x))
-          }
-        } ++ news
+        //for news values, don't keep empty, that make no sense
+        val news = (values -- existings).collect { case(k, (v, p, m)) if(v != JString("")) => NodeProperty(k,v,p,m) }
+
+        //for old value, update only according to rights
+        for {
+          updated <- sequence(props) { case prop@NodeProperty(name, value, provider, mode)  =>
```
Comment (VinceMacBuche) [inline rudder-core/src/main/scala/com/normation/rudder/domain/nodes/Node.scala]: provider here overrides provider from function declaration in the following scope

File: rudder-web/src/main/scala/bootstrap/liftweb/AppConfig.scala (hunk 8 above)
Comment (fanf) [inline rudder-web/src/main/scala/bootstrap/liftweb/AppConfig.scala]: Yeah, me to...

**Comment by fanf** [comment]
PR rebased

**Comment by fanf** [comment]
PR rebased

**Comment by fanf** [comment]
PR rebased

**Pull request merged**